source/jsonParsing.ml

```diff
@@ -64,3 +64,201 @@ let optional_list_member ~f name json =
+    end_column: int;
+  }
+  [@@deriving show, eq]
+
+  let from_decoded_range range =
+    let (start_line, end_line), (start_column, end_column) = range in
+    { start_line; end_line; start_column; end_column }
+end
+
+module Node = struct
+  type 'a t = {
+    location: Location.t option;
+    value: 'a;
+  }
+  [@@deriving show, eq]
+end
+
+exception
+  JsonmException of {
+    message: string;
+    location: Location.t;
+  }
+
+module Json = struct
+  type expression =
+    [ `Null
+    | `Bool of bool
+    | `String of string
+    | `Float of float
+    | `Int of int
+    | `List of t list
+    | `Assoc of (string * t) list
+    ]
+
+  and t = expression Node.t [@@deriving show, eq]
+
+  exception
+    TypeError of {
+      message: string;
+      json: t;
+    }
+
+  let null_node = { Node.location = None; value = `Null }
+
+  let json_of_string input =
+    let dec d =
+      match Jsonm.decode d with
+      | `Lexeme l -> l
+      | `Error error ->
+          raise
+            (JsonmException
+               {
+                 message = Format.asprintf "%a" Jsonm.pp_error error;
+                 location = Location.from_decoded_range (Jsonm.decoded_range d);
+               })
+      | `End
+      | `Await ->
+          raise
+            (JsonmException
+               {
+                 message = "Encountered end of file or input";
+                 location = Location.from_decoded_range (Jsonm.decoded_range d);
+               })
+    in
+    let rec value v k d =
+      let location = Location.from_decoded_range (Jsonm.decoded_range d) in
+      match v with
+      | `Os -> obj [] k d
+      | `As -> arr [] k d
+      | (`Null | `Bool _ | `String _ | `Float _) as v ->
+          k { Node.location = Some location; value = v } d
+      | _ ->
+          raise
+            (JsonmException
+               {
+                 message = "Encountered unexpected token or element";
+                 location = Location.from_decoded_range (Jsonm.decoded_range d);
+               })
+    and arr vs k d =
+      match dec d with
+      | `Ae ->
+          k
+            {
+              Node.location = Some (Location.from_decoded_range (Jsonm.decoded_range d));
+              value = `List (List.rev vs);
+            }
+            d
+      | v -> value v (fun v -> arr (v :: vs) k) d
+    and obj ms k d =
+      match dec d with
+      | `Oe ->
+          k
+            {
+              Node.location = Some (Location.from_decoded_range (Jsonm.decoded_range d));
+              value = `Assoc (List.rev ms);
+            }
+            d
+      | `Name n -> value (dec d) (fun v -> obj ((n, v) :: ms) k) d
+      | _ ->
+          raise
+            (JsonmException
+               {
+                 message = "Encountered unexpected token or element";
+                 location = Location.from_decoded_range (Jsonm.decoded_range d);
+               })
+    in
+    let d = Jsonm.decoder (`String input) in
+    value (dec d) (fun v _ -> v) d
```

Comment:
If possible, could we use better variable names and function names? I'm not sure what is going on here.

Reply:
token based parsing -> I took in huge inspiration from the json docs to come up with this, sorry for keeping many variable names in their format, changed it. Let me know if you think we need comments here though :)